src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,673 @@
+		LOG_ERR("FFmpeg: Failed to copy video codec parameters to stream");
+		return false;
+	}
+
+	if (avio_open(&format_context->pb, output_file_path.c_str(), AVIO_FLAG_WRITE) <
+	    0) {
+		LOG_ERR("FFmpeg: Failed to create capture file '%s'",
+		        output_file_path.c_str());
+		return false;
+	}
+
+	if (avformat_write_header(format_context, nullptr) < 0) {
+		LOG_ERR("FFmpeg: Failed to write header");
+		return false;
+	}
+
+	assert(video_stream == format_context->streams[MuxerVideoStreamIndex]);
+	assert(audio_stream == format_context->streams[MuxerAudioStreamIndex]);
+
+	return true;
+}
+
+void FFmpegMuxer::free()
+{
+	if (format_context) {
+		avio_close(format_context->pb);
+		format_context->pb = nullptr;
+		avformat_free_context(format_context);
+		format_context = nullptr;
+	}
+}
+
+bool FFmpegAudioEncoder::init(const uint32_t sample_rate)
+{
+	codec = avcodec_find_encoder(AV_CODEC_ID_AAC);
+	if (!codec) {
+		LOG_ERR("FFmpeg: Failed to find audio codec");
+		return false;
+	}
+
+	codec_context = avcodec_alloc_context3(codec);
+	if (!codec_context) {
+		LOG_ERR("FFmpeg: Failed to allocate audio context");
+		return false;
+	}
+
+	codec_context->sample_fmt  = AV_SAMPLE_FMT_FLTP;
+	codec_context->sample_rate = sample_rate;
+
+	AVChannelLayout channel_layout AV_CHANNEL_LAYOUT_STEREO;
+	if (av_channel_layout_copy(&codec_context->ch_layout, &channel_layout) < 0) {
+		LOG_ERR("FFmpeg: Failed to copy audio channel layout to codec contex");
+		return false;
+	}
+
+	if (avcodec_open2(codec_context, codec, nullptr) < 0) {
+		LOG_ERR("FFmpeg: Failed to open audio context");
+		return false;
+	}
+
+	frame = av_frame_alloc();
+	if (!frame) {
+		LOG_ERR("FFmpeg: Failed to allocate audio frame");
+		return false;
+	}
+	frame->format      = codec_context->sample_fmt;
+	frame->nb_samples  = codec_context->frame_size;
+	frame->sample_rate = sample_rate;
+	frame->pts         = 0;
+	if (av_channel_layout_copy(&frame->ch_layout, &channel_layout) < 0) {
+		LOG_ERR("FFmpeg: Failed to copy audio channel layout to frame");
+		return false;
+	}
+	if (av_frame_get_buffer(frame, 0) < 0) {
+		LOG_ERR("FFmpeg: Failed to get audio frame buffer");
+		return false;
+	}
+
+	// Resampler used to convert from interleaved 16 bit signed int
+	// To planar floating point format
+	// Can resample the audio rate in the same function call if needed
+	// (currently no resampling) I believe there is an extra step or two
+	// required for actual resampling (converting timing and such)
+	if (swr_alloc_set_opts2(&resampler_context,
+	                        &channel_layout,
+	                        AV_SAMPLE_FMT_FLTP,
+	                        sample_rate,
+	                        &channel_layout,
+	                        AV_SAMPLE_FMT_S16,
+	                        sample_rate,
+	                        0,
+	                        nullptr) < 0) {
+		LOG_ERR("FFmpeg: Failed to allocate audio resampler");
+		return false;
+	}
```

Comment:
Just wondering if we could bypass the float (audio device) -> int16 (mixer) -> float again here roundtrip, and pass floats directly from the mixed. Would be probably a more significant rework, though.

Reply:
Right, now I remember 😄 Probably the most practical thing is to leave this as is for now. I'll create a refactoring ticket for using the [-1.0, 1.0] float range all throughout.

Reply:
Yeah, this is out of scope for this PR.